diff --git a/gen-cocoa/orig/asdf.h b/gen-cocoa/no_clients/asdf.h
index 8cef031..cf72673 100644
--- a/gen-cocoa/orig/asdf.h
+++ b/gen-cocoa/no_clients/asdf.h
@@ -14,6 +14,9 @@
 #import "TProcessor.h"
 #import "TObjective-C.h"
 #import "TBase.h"
+#import "TAsyncTransport.h"
+#import "TProtocolFactory.h"
+#import "TBaseClient.h"
 
 
 @interface asdfAuthenticationEnvelope : NSObject <TBase, NSCoding> {
@@ -90,11 +93,7 @@
 - (int32_t) ping: (asdfAuthenticationEnvelope *) envelope;  // throws TException
 @end
 
-@interface asdfAuthenticatedServiceClient : NSObject <asdfAuthenticatedService> {
-  id <TProtocol> inProtocol;
-  id <TProtocol> outProtocol;
-}
-- (id) initWithProtocol: (id <TProtocol>) protocol;
+@interface asdfAuthenticatedServiceClient : TBaseClient <asdfAuthenticatedService> - (id) initWithProtocol: (id <TProtocol>) protocol;
 - (id) initWithInProtocol: (id <TProtocol>) inProtocol outProtocol: (id <TProtocol>) outProtocol;
 @end
 
@@ -111,11 +110,7 @@
 - (int32_t) ping12: (asdfAuthenticationEnvelope *) envelope;  // throws TException
 @end
 
-@interface asdfAuthenticatedService1Client : NSObject <asdfAuthenticatedService1> {
-id <TProtocol> inProtocol;
-id <TProtocol> outProtocol;
-}
-- (id) initWithProtocol: (id <TProtocol>) protocol;
+@interface asdfAuthenticatedService1Client : TBaseClient <asdfAuthenticatedService1> - (id) initWithProtocol: (id <TProtocol>) protocol;
 - (id) initWithInProtocol: (id <TProtocol>) inProtocol outProtocol: (id <TProtocol>) outProtocol;
 @end
 
@@ -133,11 +128,7 @@ NSDictionary * mMethodMap;
 - (int32_t) ping23: (asdfAuthenticationEnvelope *) envelope;  // throws TException
 @end
 
-@interface asdfAuthenticatedService2Client : NSObject <asdfAuthenticatedService2> {
-id <TProtocol> inProtocol;
-id <TProtocol> outProtocol;
-}
-- (id) initWithProtocol: (id <TProtocol>) protocol;
+@interface asdfAuthenticatedService2Client : TBaseClient <asdfAuthenticatedService2> - (id) initWithProtocol: (id <TProtocol>) protocol;
 - (id) initWithInProtocol: (id <TProtocol>) inProtocol outProtocol: (id <TProtocol>) outProtocol;
 @end
 
