diff --git a/gen-cocoa/orig/asdf.m b/gen-cocoa/no_clients/asdf.m
index 477424e..b5cbbcb 100644
--- a/gen-cocoa/orig/asdf.m
+++ b/gen-cocoa/no_clients/asdf.m
@@ -14,6 +14,9 @@
 #import "TProcessor.h"
 #import "TObjective-C.h"
 #import "TBase.h"
+#import "TAsyncTransport.h"
+#import "TProtocolFactory.h"
+#import "TBaseClient.h"
 
 
 #import "asdf.h"
@@ -815,13 +818,6 @@
   return self;
 }
 
-- (void) dealloc
-{
-  [inProtocol release_stub];
-  [outProtocol release_stub];
-  [super dealloc_stub];
-}
-
 - (void) send_ping: (asdfAuthenticationEnvelope *) envelope
 {
   [outProtocol writeMessageBeginWithName: @"ping" type: TMessageType_CALL sequenceID: 0];
@@ -834,16 +830,12 @@
   [outProtocol writeFieldStop];
   [outProtocol writeStructEnd];
   [outProtocol writeMessageEnd];
-  [[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping
 {
-  int msgType = 0;
-  [inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-  if (msgType == TMessageType_EXCEPTION) {
-    TApplicationException * x = [TApplicationException read: inProtocol];
-    [inProtocol readMessageEnd];
+  TApplicationException * x = [self checkIncomingMessageException];
+  if (x != nil)  {
     @throw x;
   }
   asdfPing_result * result = [[[asdfPing_result alloc] init] autorelease_stub];
@@ -859,6 +851,7 @@
 - (int32_t) ping: (asdfAuthenticationEnvelope *) envelope
 {
   [self send_ping : envelope];
+  [[outProtocol transport] flush];
   return [self recv_ping];
 }
 
@@ -1663,13 +1656,6 @@ outProtocol = [anOutProtocol retain_stub];
 return self;
 }
 
-- (void) dealloc
-{
-[inProtocol release_stub];
-[outProtocol release_stub];
-[super dealloc_stub];
-}
-
 - (void) send_ping11: (asdfAuthenticationEnvelope *) envelope
 {
 [outProtocol writeMessageBeginWithName: @"ping11" type: TMessageType_CALL sequenceID: 0];
@@ -1682,16 +1668,12 @@ if (envelope != nil){
 [outProtocol writeFieldStop];
 [outProtocol writeStructEnd];
 [outProtocol writeMessageEnd];
-[[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping11
 {
-int msgType = 0;
-[inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-if (msgType == TMessageType_EXCEPTION) {
-  TApplicationException * x = [TApplicationException read: inProtocol];
-  [inProtocol readMessageEnd];
+TApplicationException * x = [self checkIncomingMessageException];
+if (x != nil){
   @throw x;
 }
 asdfPing11_result * result = [[[asdfPing11_result alloc] init] autorelease_stub];
@@ -1707,6 +1689,7 @@ if ([result successIsSet]) {
 - (int32_t) ping11: (asdfAuthenticationEnvelope *) envelope
 {
 [self send_ping11 : envelope];
+[[outProtocol transport] flush];
 return [self recv_ping11];
 }
 
@@ -1722,16 +1705,12 @@ if (envelope != nil){
 [outProtocol writeFieldStop];
 [outProtocol writeStructEnd];
 [outProtocol writeMessageEnd];
-[[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping12
 {
-int msgType = 0;
-[inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-if (msgType == TMessageType_EXCEPTION) {
-  TApplicationException * x = [TApplicationException read: inProtocol];
-  [inProtocol readMessageEnd];
+TApplicationException * x = [self checkIncomingMessageException];
+if (x != nil){
   @throw x;
 }
 asdfPing12_result * result = [[[asdfPing12_result alloc] init] autorelease_stub];
@@ -1747,6 +1726,7 @@ if ([result successIsSet]) {
 - (int32_t) ping12: (asdfAuthenticationEnvelope *) envelope
 {
 [self send_ping12 : envelope];
+[[outProtocol transport] flush];
 return [self recv_ping12];
 }
 
@@ -2926,13 +2906,6 @@ outProtocol = [anOutProtocol retain_stub];
 return self;
 }
 
-- (void) dealloc
-{
-[inProtocol release_stub];
-[outProtocol release_stub];
-[super dealloc_stub];
-}
-
 - (void) send_ping21: (asdfAuthenticationEnvelope *) envelope
 {
 [outProtocol writeMessageBeginWithName: @"ping21" type: TMessageType_CALL sequenceID: 0];
@@ -2945,17 +2918,13 @@ if (envelope != nil){
 [outProtocol writeFieldStop];
 [outProtocol writeStructEnd];
 [outProtocol writeMessageEnd];
-[[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping21
 {
-int msgType = 0;
-[inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-if (msgType == TMessageType_EXCEPTION) {
-  TApplicationException * x = [TApplicationException read: inProtocol];
-  [inProtocol readMessageEnd];
-  @throw x;
+TApplicationException * x = [self checkIncomingMessageException];
+if (x != nil){
+@throw x;
 }
 asdfPing21_result * result = [[[asdfPing21_result alloc] init] autorelease_stub];
 [result read: inProtocol];
@@ -2970,6 +2939,7 @@ if ([result successIsSet]) {
 - (int32_t) ping21: (asdfAuthenticationEnvelope *) envelope
 {
 [self send_ping21 : envelope];
+[[outProtocol transport] flush];
 return [self recv_ping21];
 }
 
@@ -2985,17 +2955,13 @@ if (envelope != nil){
 [outProtocol writeFieldStop];
 [outProtocol writeStructEnd];
 [outProtocol writeMessageEnd];
-[[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping22
 {
-int msgType = 0;
-[inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-if (msgType == TMessageType_EXCEPTION) {
-  TApplicationException * x = [TApplicationException read: inProtocol];
-  [inProtocol readMessageEnd];
-  @throw x;
+TApplicationException * x = [self checkIncomingMessageException];
+if (x != nil){
+@throw x;
 }
 asdfPing22_result * result = [[[asdfPing22_result alloc] init] autorelease_stub];
 [result read: inProtocol];
@@ -3010,6 +2976,7 @@ if ([result successIsSet]) {
 - (int32_t) ping22: (asdfAuthenticationEnvelope *) envelope
 {
 [self send_ping22 : envelope];
+[[outProtocol transport] flush];
 return [self recv_ping22];
 }
 
@@ -3025,17 +2992,13 @@ if (envelope != nil){
 [outProtocol writeFieldStop];
 [outProtocol writeStructEnd];
 [outProtocol writeMessageEnd];
-[[outProtocol transport] flush];
 }
 
 - (int32_t) recv_ping23
 {
-int msgType = 0;
-[inProtocol readMessageBeginReturningName: nil type: &msgType sequenceID: NULL];
-if (msgType == TMessageType_EXCEPTION) {
-  TApplicationException * x = [TApplicationException read: inProtocol];
-  [inProtocol readMessageEnd];
-  @throw x;
+TApplicationException * x = [self checkIncomingMessageException];
+if (x != nil){
+@throw x;
 }
 asdfPing23_result * result = [[[asdfPing23_result alloc] init] autorelease_stub];
 [result read: inProtocol];
@@ -3050,6 +3013,7 @@ if ([result successIsSet]) {
 - (int32_t) ping23: (asdfAuthenticationEnvelope *) envelope
 {
 [self send_ping23 : envelope];
+[[outProtocol transport] flush];
 return [self recv_ping23];
 }
 
