diff --git a/gen-cocoa/asdf.h b/gen-cocoa/no_clients/asdf.h
index 89039e1..cf72673 100644
--- a/gen-cocoa/asdf.h
+++ b/gen-cocoa/no_clients/asdf.h
@@ -105,16 +105,6 @@
 - (id<asdfAuthenticatedService>) service;
 @end
 
-@protocol asdfAuthenticatedServiceAsync <NSObject>
-- (void) ping: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-@end
-
-@interface asdfAuthenticatedServiceClientAsync : TBaseClient <asdfAuthenticatedServiceAsync> {
-id <TAsyncTransport> asyncTransport;
-}
-- (id) initWithProtocolFactory: (id <TProtocolFactory>) factory transport: (id <TAsyncTransport>) transport;
-@end
-
 @protocol asdfAuthenticatedService1 <NSObject>
 - (int32_t) ping11: (asdfAuthenticationEnvelope *) envelope;  // throws TException
 - (int32_t) ping12: (asdfAuthenticationEnvelope *) envelope;  // throws TException
@@ -132,17 +122,6 @@ NSDictionary * mMethodMap;
 - (id<asdfAuthenticatedService1>) service;
 @end
 
-@protocol asdfAuthenticatedService1Async <NSObject>
-- (void) ping11: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-- (void) ping12: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-@end
-
-@interface asdfAuthenticatedService1ClientAsync : TBaseClient <asdfAuthenticatedService1Async> {
-id <TAsyncTransport> asyncTransport;
-}
-- (id) initWithProtocolFactory: (id <TProtocolFactory>) factory transport: (id <TAsyncTransport>) transport;
-@end
-
 @protocol asdfAuthenticatedService2 <NSObject>
 - (int32_t) ping21: (asdfAuthenticationEnvelope *) envelope;  // throws TException
 - (int32_t) ping22: (asdfAuthenticationEnvelope *) envelope;  // throws TException
@@ -161,18 +140,6 @@ NSDictionary * mMethodMap;
 - (id<asdfAuthenticatedService2>) service;
 @end
 
-@protocol asdfAuthenticatedService2Async <NSObject>
-- (void) ping21: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-- (void) ping22: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-- (void) ping23: (asdfAuthenticationEnvelope *) envelope response: (void (^)(int32_t)) responseBlock failure : (TAsyncFailureBlock) failureBlock;
-@end
-
-@interface asdfAuthenticatedService2ClientAsync : TBaseClient <asdfAuthenticatedService2Async> {
-id <TAsyncTransport> asyncTransport;
-}
-- (id) initWithProtocolFactory: (id <TProtocolFactory>) factory transport: (id <TAsyncTransport>) transport;
-@end
-
 @interface asdfasdfConstants : NSObject {
 }
 @end
